include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,384 @@
+
+  // Create first Lanczos vector if necessary
+  if (lanc_vectors.size() < 1)
+    lanc_vectors.push_back(initial_guess);
+
+  std::vector<double> t_maindiag;
+  std::vector<double> t_offdiag;
+
+  std::vector<double>
+      evecs_tridiag; // eigenvectors of tridiagonal matrix, stored in flat array
+
+  // Lanczos iteration loop
+  int it = 1;
+  for (int it_prev_check = 0; it <= _maxit; ++it)
+  {
+    // Normalize lanczos vector
+    assert(beta != 0); // TODO: set up better check for near-zero
+    lanc_vectors[it - 1] /= beta;
+
+    if (lanc_vectors.size() < static_cast<size_t>(it + 1))
+    {
+      // Add new Lanczos vector
+      lanc_vectors.push_back(VectorType(n));
+    }
+
+    // Apply operator.
+    op.vmult(lanc_vectors[it], lanc_vectors[it - 1]);
+
+    // Compute, apply, save Lanczos coefficients
+    if (it != 1)
+    {
+      lanc_vectors[it].add(-beta, lanc_vectors[it - 2]);
+      t_offdiag.push_back(beta);
+    }
+
+    alpha = lanc_vectors[it - 1] * lanc_vectors[it]; // = tridiag_{it,it}
+
+    t_maindiag.push_back(alpha);
+
+    lanc_vectors[it].add(-alpha, lanc_vectors[it - 1]);
+
+    beta = lanc_vectors[it].l2_norm(); // = tridiag_{it+1,it}
+
+    // Check convergence if requested
+    // NOTE: an alternative here for p > 0 is
+    // int((100./p)*ln(it)) > int((100./p)*ln(it-1))
+    const bool do_check =
+        it == 1 || it == _maxit ||
+        (100 * (it - it_prev_check) > _percent_overshoot * it_prev_check);
+    if (do_check)
+    {
+      // Calculate eigenpairs of tridiagonal matrix for convvergence test or at
+      // last iteration
+      details_calc_tridiag_epairs(t_maindiag, t_offdiag, num_requested, evals,
+                                  evecs_tridiag);
+
+      if (details_check_convergence(beta, num_requested, _tol, evecs_tridiag))
+      {
+        break;
+      }
+
+      // Record iteration number when this check done
+      it_prev_check = it;
+    }
+  }
+  assert(it >= num_requested);
+
+  // Calculate full operator eigenvectors from tridiagonal eigenvectors.
+  // ISSUE: may be needed to modify this code to not save all Lanczos vectors
+  // but instead recalculate them for this use.
+  // However this may be dangerous if the second Lanczos iteration
+  // has different roundoff characteristics, e.g., due to order of
+  // operation differences.
+  // ISSUE: we have not taken precautions here with regard to
+  // potential impacts of loss of orthogonality of Lanczos vectors.
+  details_calc_evecs(num_requested, it, lanc_vectors, evecs_tridiag, evecs);
+}
+
+/// \brief Lanczos solver: calculate eigenpairs from tridiagonal of Lanczos
+/// coefficients
+template <typename OperatorType, typename VectorType>
+void Lanczos<OperatorType, VectorType>::details_calc_tridiag_epairs(
+    std::vector<double> const &t_maindiag, std::vector<double> const &t_offdiag,
+    const int num_requested, std::vector<double> &evals,
+    std::vector<double> &evecs) const
```

Comment:
Looks like this one can be static too.

Readability: `evals` and `evecs` are out parameters so return them instead of passing them by reference.